src/codec/decode.go

```diff
@@ -39,149 +38,238 @@ func init() {
 	fixed64Types[FieldType_DOUBLE] = true
 }
 
-func (cb *Buffer) decodeVarintSlow() (x uint64, err error) {
-	i := cb.index
-	l := len(cb.buf)
-
-	for shift := uint(0); shift < 64; shift += 7 {
-		if i >= l {
-			err = io.ErrUnexpectedEOF
-			return
-		}
-		b := cb.buf[i]
-		i++
-		x |= (uint64(b) & 0x7F) << shift
-		if b < 0x80 {
-			cb.index = i
-			return
-		}
-	}
-
-	// The number is too large to represent in a 64-bit value.
-	err = ErrOverflow
-	return
-}
-
 // DecodeVarint reads a varint-encoded integer from the Buffer.
 // This is the format for the
 // int32, int64, uint32, uint64, bool, and enum
 // protocol buffer types.
+//
+// This implementation is inlined from https://github.com/dennwc/varint to avoid the call-site overhead
 func (cb *Buffer) DecodeVarint() (uint64, error) {
-	i := cb.index
-	buf := cb.buf
-
-	if i >= len(buf) {
+	if cb.len == 0 {
 		return 0, io.ErrUnexpectedEOF
-	} else if buf[i] < 0x80 {
-		cb.index++
-		return uint64(buf[i]), nil
-	} else if len(buf)-i < 10 {
-		return cb.decodeVarintSlow()
 	}
-
-	var b uint64
-	// we already checked the first byte
-	x := uint64(buf[i]) - 0x80
-	i++
-
-	b = uint64(buf[i])
-	i++
-	x += b << 7
-	if b&0x80 == 0 {
-		goto done
+	const (
+		step = 7
+		bit  = 1 << 7
+		mask = bit - 1
+	)
+	if cb.len >= 10 { // no bound checks
```

Comment:
`// no bound checks`

Is this actually true when you're not calling `len(cb.buf)` directly?

Reply:
I guess we could try adding:

```
_ = cb.buf[:10]
```

as a hint

Reply:
I see this is from the original library. I guess it doesn't really matter if we're still getting the perf increase we need.

Reply:
Yes, this is an artifact of the transliteration I did from the original library.

>  I guess we could try adding:
> `_ = cb.buf[:10]`

Unfortunately not as the buffer could indeed be sized smaller

Reply:
OK, I did figure out how to eliminate the bounds checks here by confirming via `go build -gcflags="-d=ssa/check_bce/debug=1"

They didn't really do much to the benchmarks though

```
ame                                         old time/op  new time/op  delta
Molecule/standard_unmarshal-8                1.64µs ±34%  1.38µs ±14%  -15.78%  (p=0.000 n=25+23)
Molecule/unmarshal_all-8                      223ns ± 6%   220ns ± 7%     ~     (p=0.303 n=24+24)
Molecule/unmarshal_loop-8                     170ns ± 6%   170ns ± 1%     ~     (p=0.155 n=21+21)
Molecule/unmarshal_single_with_molecule-8     166ns ± 4%   166ns ± 3%     ~     (p=0.063 n=23+22)
Molecule/unmarshal_multiple_with_molecule-8   993ns ± 6%   947ns ± 5%   -4.63%  (p=0.000 n=24+23)
```

Thoughts?  Commit the changes, leave as is or clean up the confusing comment?

Reply:
I'd say just remove the comment.